### PR TITLE
Support for new MinInt agent card

### PR DESCRIPTION
**Pull request by af-anssi**
This card uses the same ATR as the existing card, but the applet installed
does not have the same AID. This card actually works exactly as the
IASECC_SAGEM.

### CODE REVIEW

File: src/libopensc/card-iasecc.c [hunk 1]
```diff
@@ -612,12 +618,20 @@ iasecc_init_mi(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
-    resp_len = sizeof(resp);
-    rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
-    LOG_TEST_RET(ctx, rv, "Could not select MI's AID");
+	resp_len = sizeof(resp);
+	rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
+	LOG_TEST_GOTO_ERR(ctx, rv, "Could not select MI's AID");
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: Please only use the `err:` label (and `goto` in general) for cleanup.

File: src/libopensc/card-iasecc.c [hunk 2]
```diff
@@ -264,6 +264,9 @@ iasecc_select_mf(struct sc_card *card, struct sc_file **file_out)
 		apdu.resplen = sizeof(apdu_resp);
 		apdu.resp = apdu_resp;
 
+		if (card->type == SC_CARD_TYPE_IASECC_MI2)
+			apdu.p2 = 0x04;
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: why is this change required if the applet only changes the AID?

File: src/libopensc/card-iasecc.c [hunk 3]
```diff
@@ -323,6 +326,9 @@ iasecc_select_aid(struct sc_card *card, struct sc_aid *aid, unsigned char *out,
 	apdu.resplen = sizeof(apdu_resp);
 	apdu.resp = apdu_resp;
 
+	if (card->type == SC_CARD_TYPE_IASECC_MI2)
+		apdu.p2 = 0x04;
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: why is this change required if the applet only changes the AID?

File: src/libopensc/card-iasecc.c [hunk 4]
```diff
@@ -612,12 +612,17 @@ iasecc_init_mi(struct sc_card *card)
...
-    resp_len = sizeof(resp);
-    rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
-    LOG_TEST_RET(ctx, rv, "Could not select MI's AID");
+	resp_len = sizeof(resp);
+	rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
 
 	rv = iasecc_mi_match(card);
-    LOG_TEST_RET(ctx, rv, "Could not match MI's AID");
+
+	if(rv) {
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: use `if (rv) {`

File: src/libopensc/card-iasecc.c [hunk 5]
```diff
@@ -612,12 +612,17 @@ iasecc_init_mi(struct sc_card *card)
...
+		card->type = SC_CARD_TYPE_IASECC_MI2;
+		rv = iasecc_init_sagem(card);
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: Could you please unify the usage of `iasecc_init_amos`, `iasecc_init_sagem` and your card? Currently, they all do the same, which should be generalized under a different name.

I think the driver is not very consistently written; `iasecc_init_oberthur` and `iasecc_init_gemalto` also shares most of the code...

File: src/libopensc/card-iasecc.c [hunk 6]
```diff
@@ -562,65 +552,29 @@ iasecc_init_amos(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
+	if (card->type == SC_CARD_TYPE_IASECC_MI)   {
+		resp_len = sizeof(resp);
+		rv = iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);
+
+		rv = iasecc_mi_match(card);
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: By convention, the match function should be used to determine the card's type. I suggest you add `iasecc_select_aid(card, &MIIASECC_AID, resp, &resp_len);` to `iasecc_mi_match()` (note that you should check `rv` when selecting the applet!). (Yes, I know that this has been a mistake of the past, but giving some love to the code today, will benefit all contributors in the future.)

File: src/libopensc/card-iasecc.c [hunk 7]
```diff
@@ -562,65 +557,26 @@ iasecc_init_amos(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
+	if (card->type == SC_CARD_TYPE_IASECC_MI)   {
+		rv = iasecc_mi_match(card);
+		if (rv)
+			card->type = SC_CARD_TYPE_IASECC_MI2;
+		else
+			LOG_FUNC_RETURN(ctx, SC_SUCCESS);
```
Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: If `iasecc_mi_match()` doesn't succeed, you should not return with `SC_SUCCESS`. I think would break all other cards that are initialized with `iasecc_init_amos()`, right?

File: src/libopensc/card-iasecc.c [hunk 8]
```diff
@@ -562,65 +557,26 @@ iasecc_init_amos(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
+	if (card->type == SC_CARD_TYPE_IASECC_MI)   {
```
Comment (af-anssi) [inline src/libopensc/card-iasecc.c]: Since the type of card is tested here, there is no issue with other cards.

Comment (frankmorgner) [inline src/libopensc/card-iasecc.c]: thanks for the clearification

**Comment by frankmorgner** [comment]
Can you check whether the "old" iasecc cards are still functioning as expected?

**Comment by af-anssi** [comment]
For the SAGEM_MI2 card, the applet actually differs a bit for the one installed on SAGEM_MI cards as it really requires P2 to be set to 0x04 when selecting the MF; that is why I reintroduced this change.

I tested both SAGEM_MI and SAGEM_MI2 cards, as well as the "simple" SAGEM IAS/ECC. All are working fine via pkcs11-tool (provisioning, encryption, signature). I have no AMOS card so I cannot test it is not impacted.

**Comment by frankmorgner** [comment]
OK, fine.
